File: Examples/CloudAnchorExample/CloudAnchorExample/ExampleViewController.m
/*
 * Copyright 2018 Google LLC. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "ExampleViewController.h"

#import "CloudAnchorManager.h"

#import <dispatch/dispatch.h>

#import <ARKit/ARKit.h>
#import <ModelIO/ModelIO.h>
#import <SceneKit/ModelIO.h>

#import <ARCore/ARCore.h>

typedef NS_ENUM(NSInteger, HelloARState) {
  HelloARStateDefault,
  HelloARStateCreatingRoom,
  HelloARStateRoomCreated,
  HelloARStateHosting,
  HelloARStateHostingFinished,
  HelloARStateEnterRoomCode,
  HelloARStateResolving,
  HelloARStateResolvingFinished,
  HelloARStateCloudAnchorManagerInitFail
};

static NSString * const kPrivacyNoticeKey = @"PrivacyNoticeAccepted";
static NSString * const kPrivacyAlertTitle = @"Experience it together";
static NSString * const kPrivacyAlertMessage = @"To power this session, Google will process visual "
    "data from your camera.";
static NSString * const kPrivacyAlertContinueButtonText = @"Start now";
static NSString * const kPrivacyAlertLearnMoreButtonText = @"Learn more";
static NSString * const kPrivacyAlertBackButtonText = @"Not now";
static NSString * const kPrivacyAlertLinkURL =
    @"https://developers.google.com/ar/data-privacy";

@interface ExampleViewController () <ARSCNViewDelegate,
                                     CloudAnchorManagerDelegate,
                                     GARSessionDelegate>

@property(nonatomic, strong) ARAnchor *arAnchor;
@property(nonatomic, strong) GARAnchor *garAnchor;
@property(nonatomic, strong) NSTimer *resolveTimer;
// Node representing resolved GARAnchor. Updated when anchor changes.
@property(nonatomic, strong) SCNNode *resolvedAnchorNode;

@property(nonatomic, assign) HelloARState state;

@property(nonatomic, strong) NSString *roomCode;
@property(nonatomic, strong) NSString *message;

// Cloud Anchor Manager manages interaction between Firebase, GARSession, and ARSession.
@property(nonatomic, strong) CloudAnchorManager *cloudAnchorManager;

@end

@implementation ExampleViewController

- (void)dealloc {
  if (_resolveTimer) {
    [_resolveTimer invalidate];
  }
}

#pragma mark - Overriding UIViewController

- (BOOL)prefersStatusBarHidden {
  return YES;
}

- (void)viewDidLoad {
  [super viewDidLoad];

  [self.messageLabel setNumberOfLines:3];

  self.cloudAnchorManager = [[CloudAnchorManager alloc] initWithARSceneView:self.sceneView];
  if (self.cloudAnchorManager == nil) {
    [self enterState:HelloARStateCloudAnchorManagerInitFail];
    return;
  }

  self.cloudAnchorManager.delegate = self;
  self.sceneView.delegate = self;

  [self enterState:HelloARStateDefault];
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
  if (touches.count < 1 || self.state != HelloARStateRoomCreated) {
    return;
  }

  UITouch *touch = [[touches allObjects] firstObject];
  CGPoint touchLocation = [touch locationInView:self.sceneView];

  NSArray *hitTestResults =
      [self.sceneView hitTest:touchLocation
                        types:ARHitTestResultTypeExistingPlaneUsingExtent |
                              ARHitTestResultTypeEstimatedHorizontalPlane];

  if (hitTestResults.count > 0) {
    ARHitTestResult *result = [hitTestResults firstObject];
    [self addAnchorWithTransform:result.worldTransform];
  }
}

#pragma mark - Anchor Hosting / Resolving

- (void)resolveAnchorWithRoomCode:(NSString *)roomCode {
  self.roomCode = roomCode;
  [self enterState:HelloARStateResolving];
  [self.cloudAnchorManager resolveAnchorWithRoomCode:roomCode completion:^(GARAnchor *anchor) {
    self.garAnchor = anchor;
  }];
}

- (void)addAnchorWithTransform:(matrix_float4x4)transform {
  self.arAnchor = [[ARAnchor alloc] initWithTransform:transform];
  [self.sceneView.session addAnchor:self.arAnchor];

  self.garAnchor = [self.cloudAnchorManager hostCloudAnchor:self.arAnchor error:nil];
  [self enterState:HelloARStateHosting];
}


# pragma mark - Actions

- (IBAction)hostButtonPressed {
  if (self.state == HelloARStateDefault) {
    __weak ExampleViewController *weakSelf = self;
    [self checkPrivacyNotice:^(BOOL shouldContinue) {
      if (shouldContinue) {
        [weakSelf enterState:HelloARStateCreatingRoom];
        [weakSelf.cloudAnchorManager createRoom];
      }
    }];
  } else {
    [self enterState:HelloARStateDefault];
  }
}

- (IBAction)resolveButtonPressed {
  if (self.state == HelloARStateDefault) {
    __weak ExampleViewController *weakSelf = self;
    [self checkPrivacyNotice:^(BOOL shouldContinue) {
      if (shouldContinue) {
        [weakSelf enterState:HelloARStateEnterRoomCode];
      }
    }];
  } else {
    [self enterState:HelloARStateDefault];
  }
}

#pragma mark - CloudAnchorManagerDelegate

- (void)cloudAnchorManager:(CloudAnchorManager *)manager createdRoom:(NSString *)roomCode {
  self.roomCode = roomCode;
  [self enterState:HelloARStateRoomCreated];
}

- (void)cloudAnchorManager:(CloudAnchorManager *)manager
    failedToCreateRoomWithError:(NSError *)error {
  [self enterState:HelloARStateDefault];
}

- (void)cloudAnchorManager:(CloudAnchorManager *)manager
            didUpdateFrame:(GARFrame *)garFrame
                     error:(NSError *)error {
  if (error) {
    self.message =
        [NSString stringWithFormat:
                      @"garFrame is returned as a nil in "
                      @"CloudAnchorManager:sessioin:session:didUPdateFrame: Error description: %@",
                      [error localizedDescription]];
    [self updateMessageLabel];
    return;
  }
  for (GARAnchor *garAnchor in garFrame.updatedAnchors) {
    if ([garAnchor isEqual:self.garAnchor] && self.resolvedAnchorNode) {
      self.resolvedAnchorNode.simdTransform = garAnchor.transform;
      self.resolvedAnchorNode.hidden = !garAnchor.hasValidTransform;
    }
  }
}

- (void)cloudAnchorManager:(CloudAnchorManager *)manager
    resolveCloudAnchorReturnNilWithError:(NSError *)error {
  self.message = [NSString
      stringWithFormat:@"Resolved Cloud Anchor returned nil"
                       @"GARSession:resolveCloudAnchorWithIdentifier Error description: %@",
                       [error localizedDescription]];
  [self updateMessageLabel];
  self.garAnchor = nil;
  [self enterState:HelloARStateResolvingFinished];
}

#pragma mark - GARSessionDelegate

- (void)session:(GARSession *)session didHostAnchor:(GARAnchor *)anchor {
  if (self.state != HelloARStateHosting || ![anchor isEqual:self.garAnchor]) {
    return;
  }
  self.garAnchor = anchor;
  [self.cloudAnchorManager updateRoom:self.roomCode withAnchor:anchor];
  [self enterState:HelloARStateHostingFinished];
}

- (void)session:(GARSession *)session didFailToHostAnchor:(GARAnchor *)anchor {
  if (self.state != HelloARStateHosting || ![anchor isEqual:self.garAnchor]) {
    return;
  }
  self.garAnchor = anchor;
  [self enterState:HelloARStateHostingFinished];
}

- (void)session:(GARSession *)session didResolveAnchor:(GARAnchor *)anchor {
  if (self.state != HelloARStateResolving || ![anchor isEqual:self.garAnchor]) {
    return;
  }

  self.garAnchor = anchor;
  self.resolvedAnchorNode = [self andyNode];
  self.resolvedAnchorNode.simdTransform = anchor.transform;
  [self.sceneView.scene.rootNode addChildNode:self.resolvedAnchorNode];
  [self enterState:HelloARStateResolvingFinished];
}

- (void)session:(GARSession *)session didFailToResolveAnchor:(GARAnchor *)anchor {
  if (self.state != HelloARStateResolving || ![anchor isEqual:self.garAnchor]) {
    return;
  }
  self.garAnchor = anchor;
  [self enterState:HelloARStateResolvingFinished];
}

# pragma mark - Helper Methods

- (BOOL)privacyNoticeAccepted {
  return [[NSUserDefaults standardUserDefaults] boolForKey:kPrivacyNoticeKey];
}

- (void)setPrivacyNoticeAccepted:(BOOL)accepted {
  [[NSUserDefaults standardUserDefaults] setBool:accepted forKey:kPrivacyNoticeKey];
}

- (void)runSession {
  ARWorldTrackingConfiguration *configuration = [[ARWorldTrackingConfiguration alloc] init];
  [configuration setWorldAlignment:ARWorldAlignmentGravity];
  [configuration setPlaneDetection:ARPlaneDetectionHorizontal];

  [self.sceneView.session runWithConfiguration:configuration];
}

- (void)checkPrivacyNotice:(void (^)(BOOL shouldContinue))completion {
  __weak ExampleViewController *weakSelf = self;
  void (^innerCompletion)(BOOL shouldContinue) = ^(BOOL shouldContinue) {
    if (shouldContinue) {
      [weakSelf setPrivacyNoticeAccepted:YES];
      [weakSelf runSession];
    }
    completion(shouldContinue);
  };

  if ([self privacyNoticeAccepted]) {
    innerCompletion(YES);
    return;
  }

  UIAlertController *alertController =
      [UIAlertController alertControllerWithTitle:kPrivacyAlertTitle
                                          message:kPrivacyAlertMessage
                                   preferredStyle:UIAlertControllerStyleAlert];
  UIAlertAction *okAction =
      [UIAlertAction actionWithTitle:kPrivacyAlertContinueButtonText
                               style:UIAlertActionStyleDefault
                             handler:^(UIAlertAction *action) {
        innerCompletion(YES);
      }];
  UIAlertAction *learnMoreAction =
      [UIAlertAction actionWithTitle:kPrivacyAlertLearnMoreButtonText
                               style:UIAlertActionStyleDefault
                             handler:^(UIAlertAction *action) {
        [[UIApplication sharedApplication] openURL:[NSURL URLWithString:kPrivacyAlertLinkURL]
                                           options:@{}
                                 completionHandler:nil];
        innerCompletion(NO);
      }];
  UIAlertAction *cancelAction =
      [UIAlertAction actionWithTitle:kPrivacyAlertBackButtonText
                               style:UIAlertActionStyleDefault
                             handler:^(UIAlertAction *action) {
        innerCompletion(NO);
      }];
  [alertController addAction:okAction];
  [alertController addAction:learnMoreAction];
  [alertController addAction:cancelAction];
  [self presentViewController:alertController animated:NO completion:^{}];
}

- (void)resolveTimerFired {
  if (self.state == HelloARStateResolving) {
    self.message = @"Still resolving the anchor. Please make sure you're looking at where the "
        "Cloud Anchor was hosted. Or, try to re-join the room.";
    [self updateMessageLabel];
  }
}

- (void)updateMessageLabel {
  [self.messageLabel setText:self.message];
  self.roomCodeLabel.text = [NSString stringWithFormat:@"Room: %@", self.roomCode];
}

- (void)toggleButton:(UIButton *)button enabled:(BOOL)enabled title:(NSString *)title {
  button.enabled = enabled;
  [button setTitle:title forState:UIControlStateNormal];
}

- (NSString *)cloudStateString:(GARCloudAnchorState)cloudState {
  switch (cloudState) {
    case GARCloudAnchorStateNone:
      return @"None";
    case GARCloudAnchorStateSuccess:
      return @"Success";
    case GARCloudAnchorStateErrorInternal:
      return @"ErrorInternal";
    case GARCloudAnchorStateTaskInProgress:
      return @"TaskInProgress";
    case GARCloudAnchorStateErrorNotAuthorized:
      return @"ErrorNotAuthorized";
    case GARCloudAnchorStateErrorResourceExhausted:
      return @"ErrorResourceExhausted";
    case GARCloudAnchorStateErrorHostingDatasetProcessingFailed:
      return @"ErrorHostingDatasetProcessingFailed";
    case GARCloudAnchorStateErrorCloudIdNotFound:
      return @"ErrorCloudIdNotFound";
    case GARCloudAnchorStateErrorResolvingSdkVersionTooNew:
      return @"ErrorResolvingSdkVersionTooNew";
    case GARCloudAnchorStateErrorResolvingSdkVersionTooOld:
      return @"ErrorResolvingSdkVersionTooOld";
    case GARCloudAnchorStateErrorHostingServiceUnavailable:
      return @"ErrorHostingServiceUnavailable";
    default:
      return @"Unknown";
  }
}

- (void)showRoomCodeDialog {
  UIAlertController *alertController =
      [UIAlertController alertControllerWithTitle:@"ENTER ROOM CODE"
                                          message:@""
                                   preferredStyle:UIAlertControllerStyleAlert];
  UIAlertAction *okAction =
      [UIAlertAction actionWithTitle:@"OK"
                               style:UIAlertActionStyleDefault
                             handler:^(UIAlertAction *action) {
                               NSString *roomCode = alertController.textFields[0].text;
                               if ([roomCode length] == 0) {
                                 [self enterState:HelloARStateDefault];
                               } else {
                                 [self resolveAnchorWithRoomCode:roomCode];
                               }
                             }];
  UIAlertAction *cancelAction =
      [UIAlertAction actionWithTitle:@"CANCEL"
                               style:UIAlertActionStyleDefault
                             handler:^(UIAlertAction *action) {
                               [self enterState:HelloARStateDefault];
                             }];
  [alertController addTextFieldWithConfigurationHandler:^(UITextField *textField) {
    textField.keyboardType = UIKeyboardTypeNumberPad;
  }];
  [alertController addAction:okAction];
  [alertController addAction:cancelAction];
  [self presentViewController:alertController animated:NO completion:^{}];
}

- (void)enterState:(HelloARState)state {
  switch (state) {
    case HelloARStateCloudAnchorManagerInitFail:
      [self toggleButton:self.hostButton enabled:NO title:@"HOST disabled"];
      [self toggleButton:self.resolveButton enabled:NO title:@"RESOLVE disabled"];
      break;
    case HelloARStateDefault:
      if (self.arAnchor) {
        [self.sceneView.session removeAnchor:self.arAnchor];
        self.arAnchor = nil;
      }
      if (self.resolvedAnchorNode) {
        [self.resolvedAnchorNode removeFromParentNode];
        self.resolvedAnchorNode = nil;
      }
      if (self.garAnchor) {
        [self.cloudAnchorManager removeAnchor:self.garAnchor];
        self.garAnchor = nil;
      }
      if (self.state == HelloARStateCreatingRoom) {
        self.message = @"Failed to create room. Tap HOST or RESOLVE to begin.";
      } else {
        self.message = @"Tap HOST or RESOLVE to begin.";
      }
      if (self.state == HelloARStateEnterRoomCode) {
        [self dismissViewControllerAnimated:NO completion:^{}];
      } else if (self.state == HelloARStateResolving) {
        [self.cloudAnchorManager stopResolvingAnchorWithRoomCode:self.roomCode];
        if (self.resolveTimer) {
          [self.resolveTimer invalidate];
          self.resolveTimer = nil;
        }
      }
      [self toggleButton:self.hostButton enabled:YES title:@"HOST"];
      [self toggleButton:self.resolveButton enabled:YES title:@"RESOLVE"];
      self.roomCode = @"";
      break;
    case HelloARStateCreatingRoom:
      self.message = @"Creating room...";
      [self toggleButton:self.hostButton enabled:NO title:@"HOST"];
      [self toggleButton:self.resolveButton enabled:NO title:@"RESOLVE"];
      break;
    case HelloARStateRoomCreated:
      self.message = @"Tap on a plane to create anchor and host.";
      [self toggleButton:self.hostButton enabled:YES title:@"CANCEL"];
      [self toggleButton:self.resolveButton enabled:NO title:@"RESOLVE"];
      break;
    case HelloARStateHosting:
      self.message = @"Hosting anchor...";
      break;
    case HelloARStateHostingFinished:
      self.message =
          [NSString stringWithFormat:@"Finished hosting: %@",
                                     [self cloudStateString:self.garAnchor.cloudState]];
      break;
    case HelloARStateEnterRoomCode:
      [self showRoomCodeDialog];
      break;
    case HelloARStateResolving: {
      [self dismissViewControllerAnimated:NO completion:^{}];
      self.message = @"Resolving anchor...";
      [self toggleButton:self.hostButton enabled:NO title:@"HOST"];
      [self toggleButton:self.resolveButton enabled:YES title:@"CANCEL"];
      __weak ExampleViewController *weakSelf = self;
      self.resolveTimer = [NSTimer scheduledTimerWithTimeInterval:10.
                                                          repeats:NO
                                                            block:^(NSTimer *timer) {
                                                                    [weakSelf resolveTimerFired];
                                                                  }];
    } break;
    case HelloARStateResolvingFinished:
      if (self.resolveTimer) {
        [self.resolveTimer invalidate];
        self.resolveTimer = nil;
      }
      self.message =
          [NSString stringWithFormat:@"Finished resolving: %@",
                                     [self cloudStateString:self.garAnchor.cloudState]];
      break;
  }
  self.state = state;
  [self updateMessageLabel];
}

// Helper method to generate an SCNNode with Android geometry.
- (SCNNode *)andyNode {
  SCNScene *scene = [SCNScene sceneNamed:@"example.scnassets/andy.scn"];
  return [[scene rootNode] childNodeWithName:@"andy" recursively:NO];
}

#pragma mark - ARSCNViewDelegate

- (nullable SCNNode *)renderer:(id<SCNSceneRenderer>)renderer
                 nodeForAnchor:(ARAnchor *)anchor {
  if ([anchor isKindOfClass:[ARPlaneAnchor class]] == NO) {
    return [self andyNode];
  } else {
    return [[SCNNode alloc] init];
  }
}

- (void)renderer:(id<SCNSceneRenderer>)renderer
      didAddNode:(SCNNode *)node
       forAnchor:(ARAnchor *)anchor {
  if ([anchor isKindOfClass:[ARPlaneAnchor class]]) {
    ARPlaneAnchor *planeAnchor = (ARPlaneAnchor *)anchor;

    CGFloat width = planeAnchor.extent.x;
    CGFloat height = planeAnchor.extent.z;
    SCNPlane *plane = [SCNPlane planeWithWidth:width height:height];

    plane.materials.firstObject.diffuse.contents =
        [UIColor colorWithRed:0.0f green:0.0f blue:1.0f alpha:0.7f];

    SCNNode *planeNode = [SCNNode nodeWithGeometry:plane];

    CGFloat x = planeAnchor.center.x;
    CGFloat y = planeAnchor.center.y;
    CGFloat z = planeAnchor.center.z;
    planeNode.position = SCNVector3Make(x, y, z);
    planeNode.eulerAngles = SCNVector3Make(-M_PI / 2, 0, 0);

    [node addChildNode:planeNode];
  }
}

- (void)renderer:(id<SCNSceneRenderer>)renderer
   didUpdateNode:(SCNNode *)node
       forAnchor:(ARAnchor *)anchor {
  if ([anchor isKindOfClass:[ARPlaneAnchor class]]) {
    ARPlaneAnchor *planeAnchor = (ARPlaneAnchor *)anchor;

    SCNNode *planeNode = node.childNodes.firstObject;
    SCNPlane *plane = (SCNPlane *)planeNode.geometry;

    CGFloat width = planeAnchor.extent.x;
    CGFloat height = planeAnchor.extent.z;
    plane.width = width;
    plane.height = height;

    CGFloat x = planeAnchor.center.x;
    CGFloat y = planeAnchor.center.y;
    CGFloat z = planeAnchor.center.z;
    planeNode.position = SCNVector3Make(x, y, z);
  }
}

- (void)renderer:(id<SCNSceneRenderer>)renderer
   didRemoveNode:(SCNNode *)node
       forAnchor:(ARAnchor *)anchor {
  if ([anchor isKindOfClass:[ARPlaneAnchor class]]) {
    SCNNode *planeNode = node.childNodes.firstObject;
    [planeNode removeFromParentNode];
  }
}

@end

File: Examples/CloudAnchorExampleSwift/CloudAnchorExampleSwift/ExampleViewController.swift
/*
 * Copyright 2018 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import UIKit
import SceneKit
import ARKit
import Firebase
import ARCore
import ModelIO

enum HelloARState {
  case `default`
  case creatingRoom
  case roomCreated
  case hosting
  case hostingFinished
  case enterRoomCode
  case resolving
  case resolvingFinished
}

class ExampleViewController: UIViewController {

  @IBOutlet var sceneView: ARSCNView!
  @IBOutlet var hostButton: UIButton!
  @IBOutlet var resolveButton: UIButton!
  @IBOutlet var roomCodeLabel: UILabel!
  @IBOutlet var messageLabel: UILabel!

  private var gSession: GARSession!

  private var firebaseReference: DatabaseReference!

  private var arAnchor: ARAnchor?
  private var garAnchor: GARAnchor?

  private var state: HelloARState = .default

  private var roomCode: String = ""
  private var message: String?

  //MARK: - Overriding UIViewController
  override var prefersStatusBarHidden: Bool {
    return true
  }

  override func viewDidLoad() {
    super.viewDidLoad()

    firebaseReference = Database.database().reference()
    sceneView.delegate = self
    sceneView.session.delegate = self
    gSession = try! GARSession(apiKey: "your-api-key", bundleIdentifier: nil)
    gSession.delegate = self
    gSession.delegateQueue = DispatchQueue.main
    enter(state: .default)
  }

  override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)

    let configuration = ARWorldTrackingConfiguration()
    configuration.worldAlignment = .gravity
    configuration.planeDetection = .horizontal

    sceneView.session.run(configuration)
  }

  override func viewWillDisappear(_ animated: Bool) {
    super.viewWillDisappear(animated)
    sceneView.session.pause()
  }

  override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
    guard let touch = touches.first, state == .roomCreated else {
      return
    }

    let touchLocation = touch.location(in: sceneView)
    let hitTestResults = sceneView.hitTest(touchLocation, types: [.existingPlane, .existingPlaneUsingExtent, .estimatedHorizontalPlane])

    if let result = hitTestResults.first {
      addAnchor(transform: result.worldTransform)
    }
  }

  // MARK: - Anchor Hosting / Resolving
  func resolveAnchor(roomCode: String) {
    self.roomCode = roomCode
    enter(state: .resolving)
    firebaseReference.child("hotspot_list").child(roomCode).observe(.value) { (snapshot) in
      DispatchQueue.main.async { [weak self] in
        guard let strongSelf = self, strongSelf.state == .resolving, strongSelf.roomCode == roomCode else { return }

        if let anchorId: String = {
          if let value = snapshot.value as? [String: Any] {
            return value["hosted_anchor_id"] as? String
          }
          return nil
          }() {
          strongSelf.firebaseReference.child("hotspot_list").child(roomCode).removeAllObservers()
          strongSelf.resolveAnchor(identifier: anchorId)
        }

      }
    }
  }

  func resolveAnchor(identifier: String) {
    // Now that we have the anchor ID from firebase, we resolve the anchor.
    // Success and failure of this call is handled by the delegate methods
    // session:didResolveAnchor and session:didFailToResolveAnchor appropriately.
    do {
      garAnchor = try gSession?.resolveCloudAnchor(withIdentifier: identifier)
    } catch {
      print(error)
    }
  }

  func addAnchor(transform: matrix_float4x4) {
    let arAnchor = ARAnchor(transform: transform)
    self.arAnchor = arAnchor
    sceneView.session.add(anchor: arAnchor)

    // To share an anchor, we call host anchor here on the ARCore session.
    // session:disHostAnchor: session:didFailToHostAnchor: will get called appropriately.
    do {
      garAnchor = try gSession.hostCloudAnchor(arAnchor)
      enter(state: .hosting)
    } catch {
      print(error)
    }
  }

  // MARK: - Actions
  @IBAction func hostButtonPressed() {
    switch state {
    case .default:
      enter(state: .creatingRoom)
      createRoom()
    default:
      enter(state: .default)
    }
  }

  @IBAction func resolveButtonPressed() {
    switch state {
    case .default:
      enter(state: .enterRoomCode)
    default:
      enter(state: .default)
    }
  }
}

extension ExampleViewController: GARSessionDelegate {
  func session(_ session: GARSession, didHostAnchor anchor: GARAnchor) {
    guard state == .hosting, anchor == garAnchor else {
      return
    }
    garAnchor = anchor
    enter(state: .hostingFinished)
    firebaseReference.child("hotspot_list").child(roomCode).child("hosted_anchor_id").setValue(anchor.cloudIdentifier)
    let timestampInteger = NSDate().timeIntervalSince1970 * 1000
    let timestamp = NSNumber(value: timestampInteger)
    firebaseReference.child("hotspot_list").child(roomCode).child("updated_at_timestamp").setValue(timestamp)
  }

  func session(_ session: GARSession, didFailToHostAnchor anchor: GARAnchor) {
    guard state == .hosting, anchor == garAnchor else {
      return
    }
    self.garAnchor = anchor
    enter(state: .hostingFinished)
  }

  func session(_ session: GARSession, didResolve anchor: GARAnchor) {
    guard state == .resolving, anchor == garAnchor else {
      return
    }
    self.garAnchor = anchor
    let arAnchor = ARAnchor(transform: anchor.transform)
    self.arAnchor = ARAnchor(transform: anchor.transform)
    sceneView.session.add(anchor: arAnchor)
    enter(state: .resolvingFinished)
  }

  func session(_ session: GARSession, didFailToResolve anchor: GARAnchor) {
    guard state == .resolving, anchor == garAnchor else {
      return
    }
    self.garAnchor = anchor
    enter(state: .resolvingFinished)
  }
}

extension ExampleViewController: ARSessionDelegate {
  func session(_ session: ARSession, didUpdate frame: ARFrame) {
    do {
      try gSession.update(frame)
    } catch {
      print(error)
    }
  }
}

//MARK: - Helper Methods
private extension ExampleViewController {
  func updateMessageLabel() {
    messageLabel.text = message
    roomCodeLabel.text = "Room: \(roomCode)"
  }

  func toggle(button: UIButton, isEnabled: Bool, title: String) {
    button.isEnabled = isEnabled
    button.setTitle(title, for: .normal)
  }

  func showRoomCodeDialog() {
    let alertController = UIAlertController.init(title: "ENTER ROOM CODE", message: "", preferredStyle: .alert)
    let okAction = UIAlertAction.init(title: "OK", style: .default) { [weak self] (action) in
      if let roomCode: String = alertController.textFields?.first?.text {
        if roomCode.isEmpty {
          self?.enter(state: .default)
        } else {
          self?.resolveAnchor(roomCode: roomCode)
        }
      }
    }

    let cancelAction = UIAlertAction.init(title: "CANCEL", style: .default) { [weak self] (action) in
      self?.enter(state: .default)
    }
    alertController.addTextField { (textField) in
      textField.keyboardType = .numberPad
    }
    alertController.addAction(okAction)
    alertController.addAction(cancelAction)
    present(alertController, animated: false, completion: nil)
  }

  func enter(state: HelloARState) {
    switch state {
    case .default:
      if let arAnchor = self.arAnchor {
        sceneView.session.remove(anchor: arAnchor)
        self.arAnchor = nil
      }

      if let garAnchor = self.garAnchor {
        gSession.remove(garAnchor)
        self.garAnchor = nil
      }

      if state == .creatingRoom {
        self.message = "Failed to create room. Tap HOST or RESOLVE to begin."
      } else {
        self.message = "Tap HOST or RESOLVE to begin."
      }

      if state == .enterRoomCode {
        dismiss(animated: false, completion: nil)
      } else if state == .resolving {
        firebaseReference.child("hotspot_list").child(roomCode).removeAllObservers()
      }
      toggle(button: hostButton, isEnabled: true, title: "HOST")
      toggle(button: resolveButton, isEnabled: true, title: "RESOLVE")
      roomCode = ""
    case .creatingRoom:
      message = "Creating room..."
      toggle(button: hostButton, isEnabled: false, title: "HOST")
      toggle(button: resolveButton, isEnabled: false, title: "RESOLVE")
    case .roomCreated:
      message = "Tap on a plane to create anchor and host."
      toggle(button: hostButton, isEnabled: true, title: "CANCEL")
      toggle(button: resolveButton, isEnabled: false, title: "RESOLVE")
    case .hosting:
      message = "Hosting anchor..."
    case .hostingFinished:
      message = "Finished hosting: \(garAnchor!.cloudState.cloudStateString)"
    case .enterRoomCode:
      showRoomCodeDialog()
    case .resolving:
      dismiss(animated: false, completion: nil)
      message = "Resolving anchor..."
      toggle(button: hostButton, isEnabled: false, title: "HOST")
      toggle(button: resolveButton, isEnabled: true, title: "CANCEL")
    case .resolvingFinished:
      message = "Finished resolving: \(garAnchor!.cloudState.cloudStateString)"
    }
    self.state = state
    updateMessageLabel()
  }

  func createRoom() {
    firebaseReference.child("last_room_code").runTransactionBlock({ [weak self] (currentData) -> TransactionResult in
      let roomNumber: NSNumber = currentData.value as? NSNumber ?? 0
      let newRoomNumber = NSNumber(value: roomNumber.intValue + 1)
      let timeStamp = NSNumber(value: NSDate().timeIntervalSince1970 * 1000)

      let room: [String: Any] = ["display_name": newRoomNumber.stringValue,
                                 "updated_at_timestamp": timeStamp]
      self?.firebaseReference.child("hotspot_list").child(newRoomNumber.stringValue).setValue(room)
      currentData.value = newRoomNumber

      return TransactionResult.success(withValue: currentData)
    }) { [weak self] (error, committed, snapshot) in
      DispatchQueue.main.async {
        if error != nil {
          self?.roomCreationFailed()
        } else {
          self?.roomCreated(roomCode: (snapshot?.value as! NSNumber).stringValue)
        }
      }
    }
  }

  func roomCreated(roomCode: String) {
    self.roomCode = roomCode
    enter(state: .roomCreated)
  }

  func roomCreationFailed() {
    enter(state: .default)
  }
}

extension ExampleViewController: ARSCNViewDelegate {
  // MARK: - ARSCNViewDelegate
  func renderer(_ renderer: SCNSceneRenderer, nodeFor anchor: ARAnchor) -> SCNNode? {
    if anchor is ARPlaneAnchor {
      return SCNNode()
    } else {
      let scene = SCNScene(named: "example.scnassets/andy.scn")
      return scene?.rootNode.childNode(withName: "andy", recursively: false)
    }
  }

  func renderer(_ renderer: SCNSceneRenderer, didAdd node: SCNNode, for anchor: ARAnchor) {
    guard let planeAnchor = anchor as? ARPlaneAnchor else {
      return
    }

    let width = CGFloat(planeAnchor.extent.x)
    let height = CGFloat(planeAnchor.extent.z)
    let plane = SCNPlane(width: width, height: height)

    plane.materials.first?.diffuse.contents = UIColor(red: 0, green: 0, blue: 1.0, alpha: 0.3)

    let planeNode = SCNNode(geometry: plane)

    planeNode.position = SCNVector3Make(planeAnchor.center.x, planeAnchor.center.y, planeAnchor.center.z)
    planeNode.eulerAngles = SCNVector3Make(-Float.pi / 2 , 0, 0)
    node.addChildNode(planeNode)
  }

  func renderer(_ renderer: SCNSceneRenderer, didUpdate node: SCNNode, for anchor: ARAnchor) {
    guard let planeAnchor = anchor as? ARPlaneAnchor else {
      return
    }

    if let planeNode = node.childNodes.first, let plane = planeNode.geometry as? SCNPlane {
      let width = CGFloat(planeAnchor.extent.x)
      let height = CGFloat(planeAnchor.extent.z)
      plane.width = width
      plane.height = height
      planeNode.position = SCNVector3Make(planeAnchor.center.x, planeAnchor.center.y, planeAnchor.center.z)
    }
  }

  func renderer(_ renderer: SCNSceneRenderer, didRemove node: SCNNode, for anchor: ARAnchor) {
    guard anchor is ARPlaneAnchor else {
      return
    }
    if let planeNode = node.childNodes.first {
      planeNode.removeFromParentNode()
    }
  }
}

private extension GARCloudAnchorState {
  var cloudStateString: String {
    switch self {
    case .none:
      return "None"
    case .success:
      return "Success"
    case .errorInternal:
      return "ErrorInternal"
    case .taskInProgress:
      return "TaskInProgress"
    case .errorNotAuthorized:
      return "ErrorNotAuthorized"
    case .errorResourceExhausted:
      return "ErrorResourceExhausted"
    case .errorServiceUnavailable:
      return "ErrorResourceExhausted"
    case .errorHostingDatasetProcessingFailed:
      return "ErrorHostingDatasetProcessingFailed"
    case .errorCloudIdNotFound:
      return "ErrorCloudIdNotFound"
    case .errorResolvingSdkVersionTooNew:
      return "ErrorResolvingSdkVersionTooNew"
    case .errorResolvingSdkVersionTooOld:
      return "ErrorResolvingSdkVersionTooOld"
    case .errorResolvingLocalizationNoMatch:
      return "ErrorResolvingLocalizationNoMatch"
    }
  }
}
